google-cloud-pubsub/lib/google/cloud/pubsub/batch_publisher.rb

```diff
@@ -26,32 +28,53 @@ module PubSub
       #   pubsub = Google::Cloud::PubSub.new
       #
       #   topic = pubsub.topic "my-topic"
-      #   msgs = topic.publish do |t|
-      #     t.publish "task 1 completed", foo: :bar
-      #     t.publish "task 2 completed", foo: :baz
-      #     t.publish "task 3 completed", foo: :bif
+      #   msgs = topic.publish do |batch_publisher|
+      #     batch_publisher.publish "task 1 completed", foo: :bar
+      #     batch_publisher.publish "task 2 completed", foo: :baz
+      #     batch_publisher.publish "task 3 completed", foo: :bif
       #   end
+      #
       class BatchPublisher
         ##
         # @private The messages to publish
         attr_reader :messages
 
         ##
         # @private Create a new instance of the object.
-        def initialize data = nil, attributes = {}
+        def initialize data, attributes, ordering_key, extra_attrs
           @messages = []
           @mode = :batch
           return if data.nil?
           @mode = :single
-          publish data, attributes
+          publish data, attributes, ordering_key: ordering_key, **extra_attrs
         end
 
         ##
         # Add a message to the batch to be published to the topic.
         # All messages added to the batch will be published at once.
         # See {Google::Cloud::PubSub::Topic#publish}
-        def publish data, attributes = {}
-          @messages << create_pubsub_message(data, attributes)
+        #
+        # @param [String, File] data The message payload. This will be converted
+        #   to bytes encoded as ASCII-8BIT.
+        # @param [Hash] attributes Optional attributes for the message.
+        # @param [String] ordering_key Identifies related messages for which
+        #   publish order should be respected.
+        #
+        # @example Multiple messages can be sent at the same time using a block:
+        #   require "google/cloud/pubsub"
+        #
+        #   pubsub = Google::Cloud::PubSub.new
+        #
+        #   topic = pubsub.topic "my-topic"
+        #   msgs = topic.publish do |batch_publisher|
+        #     batch_publisher.publish "task 1 completed", foo: :bar
+        #     batch_publisher.publish "task 2 completed", foo: :baz
+        #     batch_publisher.publish "task 3 completed", foo: :bif
+        #   end
+        #
+        def publish data, attributes = nil, ordering_key: nil, **extra_attrs
```

Comment:
This will be breaking for anyone who was using the earlier interface like this:

```ruby
batch_publisher.publish "hello", my_attr: "foo"
```

Are we okay with this?

Reply:
I didn't think that the technique using `**extra_attrs` would introduce a breaking change. It is based on [the original `ordering_key` commit](https://github.com/googleapis/google-cloud-ruby/commit/4300bebf5acbb2295cb1668f6563a72a717958f8), which changed:

```
def publish_async data = nil, attributes = {}, &block
```

to 

```
def publish_async data = nil, attributes = nil, ordering_key: nil, **extra_attrs, &callback
```

This PR changes:

```
def publish data = nil, attributes = {}
```

to 

```
def publish data = nil, attributes = nil, ordering_key: nil, **extra_attrs, &block
```

Existing regression tests such as [/pubsub_test.rb#L75-L95](https://github.com/googleapis/google-cloud-ruby/blob/a7681bc065ada563049edc524a25f1324e53d059/google-cloud-pubsub/acceptance/pubsub/pubsub_test.rb#L75-L95) (acceptance) and [publish_test.rb#L86-L102](https://github.com/googleapis/google-cloud-ruby/blob/a7681bc065ada563049edc524a25f1324e53d059/google-cloud-pubsub/test/google/cloud/pubsub/topic/publish_test.rb#L86-L102) (unit) are still passing in this PR.

Reply:
Oh, I see, `Convert.pubsub_message` combines `attributes` and `extra_attrs` to compensate.

Reply:
So the only potentially breaking case is if someone is already passing `ordering_key:` and expecting it to be an attribute. That's probably unlikely enough for us to ignore.